# **CREDIT RISK SCORING SYSTEM**

#### Problem Statement: Developing a Robust Credit Risk Scoring System using Machine Learning

In the rapidly evolving fintech industry, accurate credit risk assessment is paramount to financial institutions. Credit risk scoring is a critical tool that helps lenders evaluate the likelihood of a borrower defaulting on a loan. Traditional credit scoring models often rely on a limited set of financial indicators and can be prone to biases and inaccuracies. With the advent of big data and advanced machine learning techniques, there is a significant opportunity to enhance the accuracy and reliability of credit risk scoring models.

#### Objective

The primary objective of this project is to develop a robust, scalable, and interpretable machine learning model that predicts the probability of default for loan applicants. The model aims to leverage a wide array of features, including demographic data, financial status, credit history, and behavioral data, to deliver accurate credit risk assessments.

#### Key Questions

1. **Data Integration**: How can we effectively integrate diverse data sources to build a comprehensive dataset for credit risk assessment?
2. **Feature Engineering**: What are the most predictive features for assessing credit risk, and how can we engineer new features to improve model performance?
3. **Model Selection**: Which machine learning algorithms provide the best performance in terms of accuracy, interpretability, and computational efficiency for credit risk scoring?
4. **Model Evaluation**: What metrics should be used to evaluate the performance of the credit risk model, and how can we ensure the model generalizes well to unseen data?
5. **Bias and Fairness**: How can we detect and mitigate biases in the credit risk model to ensure fair treatment of all applicants?
6. **Deployment and Monitoring**: How can we deploy the model in a real-world setting, and what mechanisms should be in place for continuous monitoring and updating of the model?

#### Scope and Deliverables

1. **Data Collection and Preprocessing**:
   - Collect and preprocess data from the chosen datasets (Home Credit Default Risk, LendingClub Loan Data, and Give Me Some Credit).
   - Handle missing values, outliers, and data inconsistencies.

2. **Exploratory Data Analysis (EDA)**:
   - Conduct EDA to understand data distribution, correlations, and key insights.
   - Visualize important patterns and relationships in the data.

3. **Feature Engineering**:
   - Develop and select features that significantly impact credit risk prediction.
   - Implement techniques such as one-hot encoding, scaling, and normalization.

4. **Model Development**:
   - Train and compare multiple machine learning models, including Logistic Regression, Decision Trees, Random Forests, Gradient Boosting Machines, XGBoost, and Neural Networks.
   - Perform hyperparameter tuning to optimize model performance.

5. **Model Evaluation**:
   - Evaluate models using metrics such as AUC-ROC, Precision-Recall, F1 Score, and Confusion Matrix.
   - Ensure model interpretability using SHAP values or LIME.

6. **Deployment**:
   - Develop an API for the model using Flask or FastAPI.
   - Deploy the model on a cloud platform (e.g., AWS, GCP, Azure) and integrate with a front-end application.

7. **Monitoring and Maintenance**:
   - Implement monitoring tools to track model performance over time.
   - Set up a pipeline for continuous learning and model updates based on new data.

#### Expected Impact

By developing a sophisticated credit risk scoring system, this project aims to:
- Enhance the accuracy and reliability of credit risk assessments.
- Enable financial institutions to make more informed lending decisions.
- Reduce the risk of defaults and financial losses.
- Promote fair and unbiased credit evaluation processes.

#### Conclusion

This project will push the limits of data science in the fintech space by leveraging advanced machine learning techniques to address a critical challenge in credit risk assessment. The outcomes will not only benefit financial institutions but also contribute to the broader goal of financial inclusion and stability.

### **IMPORT NECESSARY MODULES**

In [4]:
# Importing standard libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

# Importing libraries for data preprocessing
from sklearn.preprocessing import StandardScaler, MinMaxScaler, LabelEncoder, OneHotEncoder
from sklearn.model_selection import train_test_split, GridSearchCV, RandomizedSearchCV
from sklearn.impute import SimpleImputer

# Importing libraries for machine learning models
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier
from xgboost import XGBClassifier
from lightgbm import LGBMClassifier
from sklearn.neural_network import MLPClassifier

# Importing libraries for model evaluation
from sklearn.metrics import roc_auc_score, accuracy_score, precision_score, recall_score, f1_score, confusion_matrix, roc_curve, precision_recall_curve

# Importing libraries for model interpretability
import shap
import lime
import lime.lime_tabular

# Importing libraries for API development and deployment
from flask import Flask, request, jsonify
import joblib

# Miscellaneous libraries
import warnings
warnings.filterwarnings('ignore')

# Display all rows and columns
pd.set_option('display.max_rows', None)
pd.set_option('display.max_columns', None)


### **IMPORT THE TRAIN AND TEST CSV FILES**

In [5]:
# Import training set
train_df = pd.read_csv(r"C:\Users\Black Concept\WorkSpace\ALTSCHOOL\Datasets\application_train.csv")

# Import test set
test_df = pd.read_csv(r"C:\Users\Black Concept\WorkSpace\ALTSCHOOL\Datasets\application_test.csv")

### **DATA PREPROCESSING**

  Cleaning: Handle missing values, outliers, and data inconsistencies.

  Normalization/Standardization: Scale numerical features.

  Encoding: Convert categorical variables into numerical values using techniques like one-hot encoding or label encoding.

  Feature Engineering: Create new features based on domain knowledge (e.g., debt-to-income ratio, credit utilization rate).

This project already broke down the dataset we are using into train and test set already which is a good approach so we can take a look at the two datasets separately during exploratory data analysis where we take a look at the data- the various connections between features and distribution of features as well.

Taking a look at the training and test dataset is very import to avoid bias which could lead to overfitting and again, its important to see that our features are well represented and distributed evenly across the two datasets.

### **How large is the dataset that we are working with and what are the different features?**

In [6]:
# Check the size of the datasets
train_size = train_df.shape # Training size
test_size = test_df.shape # Test size

print(f'The training set has {train_size[0]} rows (observations) and {train_size[-1]} columns (features)')

print(f'The test set has {test_size[0]} rows (observations) and {test_size[-1]} columns (features)')

The training set has 307511 rows (observations) and 122 columns (features)
The test set has 48744 rows (observations) and 121 columns (features)


In [7]:
# Check the different features(columns) and their respective descriptions
columns = list(train_df.columns)
print(columns)

['SK_ID_CURR', 'TARGET', 'NAME_CONTRACT_TYPE', 'CODE_GENDER', 'FLAG_OWN_CAR', 'FLAG_OWN_REALTY', 'CNT_CHILDREN', 'AMT_INCOME_TOTAL', 'AMT_CREDIT', 'AMT_ANNUITY', 'AMT_GOODS_PRICE', 'NAME_TYPE_SUITE', 'NAME_INCOME_TYPE', 'NAME_EDUCATION_TYPE', 'NAME_FAMILY_STATUS', 'NAME_HOUSING_TYPE', 'REGION_POPULATION_RELATIVE', 'DAYS_BIRTH', 'DAYS_EMPLOYED', 'DAYS_REGISTRATION', 'DAYS_ID_PUBLISH', 'OWN_CAR_AGE', 'FLAG_MOBIL', 'FLAG_EMP_PHONE', 'FLAG_WORK_PHONE', 'FLAG_CONT_MOBILE', 'FLAG_PHONE', 'FLAG_EMAIL', 'OCCUPATION_TYPE', 'CNT_FAM_MEMBERS', 'REGION_RATING_CLIENT', 'REGION_RATING_CLIENT_W_CITY', 'WEEKDAY_APPR_PROCESS_START', 'HOUR_APPR_PROCESS_START', 'REG_REGION_NOT_LIVE_REGION', 'REG_REGION_NOT_WORK_REGION', 'LIVE_REGION_NOT_WORK_REGION', 'REG_CITY_NOT_LIVE_CITY', 'REG_CITY_NOT_WORK_CITY', 'LIVE_CITY_NOT_WORK_CITY', 'ORGANIZATION_TYPE', 'EXT_SOURCE_1', 'EXT_SOURCE_2', 'EXT_SOURCE_3', 'APARTMENTS_AVG', 'BASEMENTAREA_AVG', 'YEARS_BEGINEXPLUATATION_AVG', 'YEARS_BUILD_AVG', 'COMMONAREA_AVG', 'ELE

Features in the dataset and their descriptions

1. **SK_ID_CURR**: Unique identifier for the client.
2. **TARGET**: Target variable (0 = repaid loan on time, 1 = defaulted on loan).
3. **NAME_CONTRACT_TYPE**: Type of loan contract (e.g., Cash loans, Revolving loans).
4. **CODE_GENDER**: Gender of the client (M = Male, F = Female).
5. **FLAG_OWN_CAR**: Flag indicating if the client owns a car (Y = Yes, N = No).
6. **FLAG_OWN_REALTY**: Flag indicating if the client owns real estate (Y = Yes, N = No).
7. **CNT_CHILDREN**: Number of children the client has.
8. **AMT_INCOME_TOTAL**: Total income of the client.
9. **AMT_CREDIT**: Credit amount of the loan.
10. **AMT_ANNUITY**: Loan annuity.
11. **AMT_GOODS_PRICE**: Price of the goods for which the loan is given.
12. **NAME_TYPE_SUITE**: Who accompanied the client when applying for the loan.
13. **NAME_INCOME_TYPE**: Income type of the client (e.g., Working, Commercial associate, Pensioner).
14. **NAME_EDUCATION_TYPE**: Education level of the client.
15. **NAME_FAMILY_STATUS**: Family status of the client.
16. **NAME_HOUSING_TYPE**: Housing situation of the client.
17. **REGION_POPULATION_RELATIVE**: Relative population of the region.
18. **DAYS_BIRTH**: Client's age in days.
19. **DAYS_EMPLOYED**: Employment duration in days.
20. **DAYS_REGISTRATION**: Registration date of the client's application.
21. **DAYS_ID_PUBLISH**: ID publish date.
22. **OWN_CAR_AGE**: Age of the client's car.
23. **FLAG_MOBIL**: Flag indicating if the client has a mobile phone (1 = Yes, 0 = No).
24. **FLAG_EMP_PHONE**: Flag indicating if the client has a work phone (1 = Yes, 0 = No).
25. **FLAG_WORK_PHONE**: Flag indicating if the client has a work phone (1 = Yes, 0 = No).
26. **FLAG_CONT_MOBILE**: Flag indicating if the client's mobile phone is permanent (1 = Yes, 0 = No).
27. **FLAG_PHONE**: Flag indicating if the client has a phone (1 = Yes, 0 = No).
28. **FLAG_EMAIL**: Flag indicating if the client has an email (1 = Yes, 0 = No).
29. **OCCUPATION_TYPE**: Occupation of the client.
30. **CNT_FAM_MEMBERS**: Number of family members.
31. **REGION_RATING_CLIENT**: Rating of the region where the client lives.
32. **REGION_RATING_CLIENT_W_CITY**: Rating of the region and city where the client lives.
33. **WEEKDAY_APPR_PROCESS_START**: Weekday when the application was processed.
34. **HOUR_APPR_PROCESS_START**: Hour when the application was processed.
35. **REG_REGION_NOT_LIVE_REGION**: Flag indicating if the client lives in a different region than the application region.
36. **REG_REGION_NOT_WORK_REGION**: Flag indicating if the client works in a different region than the application region.
37. **LIVE_REGION_NOT_WORK_REGION**: Flag indicating if the client lives and works in different regions.
38. **REG_CITY_NOT_LIVE_CITY**: Flag indicating if the client lives in a different city than the application city.
39. **REG_CITY_NOT_WORK_CITY**: Flag indicating if the client works in a different city than the application city.
40. **LIVE_CITY_NOT_WORK_CITY**: Flag indicating if the client lives and works in different cities.
41. **ORGANIZATION_TYPE**: Type of organization where the client works.
42. **EXT_SOURCE_1**: Normalized score from an external data source.
43. **EXT_SOURCE_2**: Normalized score from an external data source.
44. **EXT_SOURCE_3**: Normalized score from an external data source.
45. **APARTMENTS_AVG**: Average size of apartments.
46. **BASEMENTAREA_AVG**: Average size of the basement area.
47. **YEARS_BEGINEXPLUATATION_AVG**: Average years of operation of the building.
48. **YEARS_BUILD_AVG**: Average year of building construction.
49. **COMMONAREA_AVG**: Average size of common area.
50. **ELEVATORS_AVG**: Average number of elevators.
51. **ENTRANCES_AVG**: Average number of entrances.
52. **FLOORSMAX_AVG**: Average number of floors in the building.
53. **FLOORSMIN_AVG**: Average number of floors in the building (minimum).
54. **LANDAREA_AVG**: Average size of the land area.
55. **LIVINGAPARTMENTS_AVG**: Average size of living apartments.
56. **LIVINGAREA_AVG**: Average size of living area.
57. **NONLIVINGAPARTMENTS_AVG**: Average size of non-living apartments.
58. **NONLIVINGAREA_AVG**: Average size of non-living area.
59. **APARTMENTS_MODE**: Mode of apartment size.
60. **BASEMENTAREA_MODE**: Mode of basement area size.
61. **YEARS_BEGINEXPLUATATION_MODE**: Mode of years of building operation.
62. **YEARS_BUILD_MODE**: Mode of building construction year.
63. **COMMONAREA_MODE**: Mode of common area size.
64. **ELEVATORS_MODE**: Mode of number of elevators.
65. **ENTRANCES_MODE**: Mode of number of entrances.
66. **FLOORSMAX_MODE**: Mode of number of floors in the building (maximum).
67. **FLOORSMIN_MODE**: Mode of number of floors in the building (minimum).
68. **LANDAREA_MODE**: Mode of land area size.
69. **LIVINGAPARTMENTS_MODE**: Mode of living apartments size.
70. **LIVINGAREA_MODE**: Mode of living area size.
71. **NONLIVINGAPARTMENTS_MODE**: Mode of non-living apartments size.
72. **NONLIVINGAREA_MODE**: Mode of non-living area size.
73. **APARTMENTS_MEDI**: Median size of apartments.
74. **BASEMENTAREA_MEDI**: Median size of the basement area.
75. **YEARS_BEGINEXPLUATATION_MEDI**: Median years of building operation.
76. **YEARS_BUILD_MEDI**: Median year of building construction.
77. **COMMONAREA_MEDI**: Median size of common area.
78. **ELEVATORS_MEDI**: Median number of elevators.
79. **ENTRANCES_MEDI**: Median number of entrances.
80. **FLOORSMAX_MEDI**: Median number of floors in the building (maximum).
81. **FLOORSMIN_MEDI**: Median number of floors in the building (minimum).
82. **LANDAREA_MEDI**: Median size of land area.
83. **LIVINGAPARTMENTS_MEDI**: Median size of living apartments.
84. **LIVINGAREA_MEDI**: Median size of living area.
85. **NONLIVINGAPARTMENTS_MEDI**: Median size of non-living apartments.
86. **NONLIVINGAREA_MEDI**: Median size of non-living area.
87. **FONDKAPREMONT_MODE**: Type of maintenance fund.
88. **HOUSETYPE_MODE**: Type of house.
89. **TOTALAREA_MODE**: Total area.
90. **WALLSMATERIAL_MODE**: Material of the walls.
91. **EMERGENCYSTATE_MODE**: Emergency state of the house.
92. **OBS_30_CNT_SOCIAL_CIRCLE**: Observed number of social circle (30 days).
93. **DEF_30_CNT_SOCIAL_CIRCLE**: Defected number of social circle (30 days).
94. **OBS_60_CNT_SOCIAL_CIRCLE**: Observed number of social circle (60 days).
95. **DEF_60_CNT_SOCIAL_CIRCLE**: Defected number of social circle (60 days).
96. **DAYS_LAST_PHONE_CHANGE**: Days since last phone change.
97. **FLAG_DOCUMENT_2**: Flag indicating if document 2 is present.
98. **FLAG_DOCUMENT_3**: Flag indicating if document 3 is present.
99. **FLAG_DOCUMENT_4**: Flag indicating if document 4 is present.
100. **FLAG_DOCUMENT_5**: Flag indicating if document 5 is present.
101. **FLAG_DOCUMENT_6**: Flag indicating if document 6 is present.
102. **FLAG_DOCUMENT_7**: Flag indicating if document 7 is present.
103. **FLAG_DOCUMENT_8**: Flag indicating if document 8 is present.
104. **FLAG_DOCUMENT_9**: Flag indicating if document 9 is present.
105. **FLAG_DOCUMENT_10**: Flag indicating if document 10 is present.
106. **FLAG_DOCUMENT_11**: Flag indicating if document 11 is present.
107. **FLAG_DOCUMENT_12**: Flag indicating if document 12 is present.
108. **FLAG_DOCUMENT_13**: Flag indicating if document 13 is present.
109. **FLAG_DOCUMENT_14**: Flag indicating if document 14 is present.
110. **FLAG_DOCUMENT_15**: Flag indicating if document 15 is present.
111. **FLAG_DOCUMENT_16**: Flag indicating if document 16 is present.
112. **FLAG_DOCUMENT_17**: Flag indicating if document 17 is present.
113. **FLAG_DOCUMENT_18**: Flag indicating if document 18 is present.
114. **FLAG_DOCUMENT_19**: Flag indicating if document 19 is present.
115. **FLAG_DOCUMENT_20**: Flag indicating if document 20 is present.
116. **FLAG_DOCUMENT_21**: Flag indicating if document 21 is present.
117. **AMT_REQ_CREDIT_BUREAU_HOUR**: Number of enquiries to Credit Bureau about the client per hour.
118. **AMT_REQ_CREDIT_BUREAU_DAY**: Number of enquiries to Credit Bureau about the client per day.
119. **AMT_REQ_CREDIT_BUREAU_WEEK**: Number of enquiries to Credit Bureau about the client per week.
120. **AMT_REQ_CREDIT_BUREAU_MON**: Number of enquiries to Credit Bureau about the client per month.
121. **AMT_REQ_CREDIT_BUREAU_QRT**: Number of enquiries to Credit Bureau about the client per quarter.
122. **AMT_REQ_CREDIT_BUREAU_YEAR**: Number of enquiries to Credit Bureau about the client per year.

### **DATA CLEANING**

    Rename the columns

Firstly, I want to rename the columns more intuitively in a way that would be easy to read and comprehend

In [8]:
# Create a dictionary for the new column names
new_column_dict = {
    'SK_ID_CURR': 'Client_ID',
    'TARGET': 'Loan_Status',
    'NAME_CONTRACT_TYPE': 'Contract_Type',
    'CODE_GENDER': 'Gender',
    'FLAG_OWN_CAR': 'Own_Car',
    'FLAG_OWN_REALTY': 'Own_Realty',
    'CNT_CHILDREN': 'Number_of_Children',
    'AMT_INCOME_TOTAL': 'Total_Income',
    'AMT_CREDIT': 'Credit_Amount',
    'AMT_ANNUITY': 'Annuity_Amount',
    'AMT_GOODS_PRICE': 'Goods_Price',
    'NAME_TYPE_SUITE': 'Accompanied_By',
    'NAME_INCOME_TYPE': 'Income_Type',
    'NAME_EDUCATION_TYPE': 'Education_Level',
    'NAME_FAMILY_STATUS': 'Family_Status',
    'NAME_HOUSING_TYPE': 'Housing_Type',
    'REGION_POPULATION_RELATIVE': 'Region_Population_Ratio',
    'DAYS_BIRTH': 'Age_Days',
    'DAYS_EMPLOYED': 'Employment_Duration_Days',
    'DAYS_REGISTRATION': 'Registration_Duration_Days',
    'DAYS_ID_PUBLISH': 'ID_Publish_Duration_Days',
    'OWN_CAR_AGE': 'Car_Age',
    'FLAG_MOBIL': 'Has_Mobile',
    'FLAG_EMP_PHONE': 'Has_Work_Phone',
    'FLAG_WORK_PHONE': 'Has_Work_Phone2',
    'FLAG_CONT_MOBILE': 'Has_Permanent_Mobile',
    'FLAG_PHONE': 'Has_Phone',
    'FLAG_EMAIL': 'Has_Email',
    'OCCUPATION_TYPE': 'Occupation',
    'CNT_FAM_MEMBERS': 'Family_Members_Count',
    'REGION_RATING_CLIENT': 'Region_Rating',
    'REGION_RATING_CLIENT_W_CITY': 'Region_City_Rating',
    'WEEKDAY_APPR_PROCESS_START': 'Application_Weekday',
    'HOUR_APPR_PROCESS_START': 'Application_Hour',
    'REG_REGION_NOT_LIVE_REGION': 'Different_Region_Living',
    'REG_REGION_NOT_WORK_REGION': 'Different_Region_Working',
    'LIVE_REGION_NOT_WORK_REGION': 'Different_Region_Living_Working',
    'REG_CITY_NOT_LIVE_CITY': 'Different_City_Living',
    'REG_CITY_NOT_WORK_CITY': 'Different_City_Working',
    'LIVE_CITY_NOT_WORK_CITY': 'Different_City_Living_Working',
    'ORGANIZATION_TYPE': 'Organization_Type',
    'EXT_SOURCE_1': 'External_Source_1',
    'EXT_SOURCE_2': 'External_Source_2',
    'EXT_SOURCE_3': 'External_Source_3',
    'APARTMENTS_AVG': 'Apartments_Avg',
    'BASEMENTAREA_AVG': 'Basement_Area_Avg',
    'YEARS_BEGINEXPLUATATION_AVG': 'Years_Begin_Expluatation_Avg',
    'YEARS_BUILD_AVG': 'Years_Build_Avg',
    'COMMONAREA_AVG': 'Common_Area_Avg',
    'ELEVATORS_AVG': 'Elevators_Avg',
    'ENTRANCES_AVG': 'Entrances_Avg',
    'FLOORSMAX_AVG': 'Floors_Max_Avg',
    'FLOORSMIN_AVG': 'Floors_Min_Avg',
    'LANDAREA_AVG': 'Land_Area_Avg',
    'LIVINGAPARTMENTS_AVG': 'Living_Apartments_Avg',
    'LIVINGAREA_AVG': 'Living_Area_Avg',
    'NONLIVINGAPARTMENTS_AVG': 'Non_Living_Apartments_Avg',
    'NONLIVINGAREA_AVG': 'Non_Living_Area_Avg',
    'APARTMENTS_MODE': 'Apartments_Mode',
    'BASEMENTAREA_MODE': 'Basement_Area_Mode',
    'YEARS_BEGINEXPLUATATION_MODE': 'Years_Begin_Expluatation_Mode',
    'YEARS_BUILD_MODE': 'Years_Build_Mode',
    'COMMONAREA_MODE': 'Common_Area_Mode',
    'ELEVATORS_MODE': 'Elevators_Mode',
    'ENTRANCES_MODE': 'Entrances_Mode',
    'FLOORSMAX_MODE': 'Floors_Max_Mode',
    'FLOORSMIN_MODE': 'Floors_Min_Mode',
    'LANDAREA_MODE': 'Land_Area_Mode',
    'LIVINGAPARTMENTS_MODE': 'Living_Apartments_Mode',
    'LIVINGAREA_MODE': 'Living_Area_Mode',
    'NONLIVINGAPARTMENTS_MODE': 'Non_Living_Apartments_Mode',
    'NONLIVINGAREA_MODE': 'Non_Living_Area_Mode',
    'APARTMENTS_MEDI': 'Apartments_Median',
    'BASEMENTAREA_MEDI': 'Basement_Area_Median',
    'YEARS_BEGINEXPLUATATION_MEDI': 'Years_Begin_Expluatation_Median',
    'YEARS_BUILD_MEDI': 'Years_Build_Median',
    'COMMONAREA_MEDI': 'Common_Area_Median',
    'ELEVATORS_MEDI': 'Elevators_Median',
    'ENTRANCES_MEDI': 'Entrances_Median',
    'FLOORSMAX_MEDI': 'Floors_Max_Median',
    'FLOORSMIN_MEDI': 'Floors_Min_Median',
    'LANDAREA_MEDI': 'Land_Area_Median',
    'LIVINGAPARTMENTS_MEDI': 'Living_Apartments_Median',
    'LIVINGAREA_MEDI': 'Living_Area_Median',
    'NONLIVINGAPARTMENTS_MEDI': 'Non_Living_Apartments_Median',
    'NONLIVINGAREA_MEDI': 'Non_Living_Area_Median',
    'FONDKAPREMONT_MODE': 'Maintenance_Fund_Mode',
    'HOUSETYPE_MODE': 'House_Type_Mode',
    'TOTALAREA_MODE': 'Total_Area_Mode',
    'WALLSMATERIAL_MODE': 'Walls_Material_Mode',
    'EMERGENCYSTATE_MODE': 'Emergency_State_Mode',
    'OBS_30_CNT_SOCIAL_CIRCLE': 'Obs_30_Social_Circle',
    'DEF_30_CNT_SOCIAL_CIRCLE': 'Def_30_Social_Circle',
    'OBS_60_CNT_SOCIAL_CIRCLE': 'Obs_60_Social_Circle',
    'DEF_60_CNT_SOCIAL_CIRCLE': 'Def_60_Social_Circle',
    'DAYS_LAST_PHONE_CHANGE': 'Days_Since_Last_Phone_Change',
    'FLAG_DOCUMENT_2': 'Has_Document_2',
    'FLAG_DOCUMENT_3': 'Has_Document_3',
    'FLAG_DOCUMENT_4': 'Has_Document_4',
    'FLAG_DOCUMENT_5': 'Has_Document_5',
    'FLAG_DOCUMENT_6': 'Has_Document_6',
    'FLAG_DOCUMENT_7': 'Has_Document_7',
    'FLAG_DOCUMENT_8': 'Has_Document_8',
    'FLAG_DOCUMENT_9': 'Has_Document_9',
    'FLAG_DOCUMENT_10': 'Has_Document_10',
    'FLAG_DOCUMENT_11': 'Has_Document_11',
    'FLAG_DOCUMENT_12': 'Has_Document_12',
    'FLAG_DOCUMENT_13': 'Has_Document_13',
    'FLAG_DOCUMENT_14': 'Has_Document_14',
    'FLAG_DOCUMENT_15': 'Has_Document_15',
    'FLAG_DOCUMENT_16': 'Has_Document_16',
    'FLAG_DOCUMENT_17': 'Has_Document_17',
    'FLAG_DOCUMENT_18': 'Has_Document_18',
    'FLAG_DOCUMENT_19': 'Has_Document_19',
    'FLAG_DOCUMENT_20': 'Has_Document_20',
    'FLAG_DOCUMENT_21': 'Has_Document_21',
    'AMT_REQ_CREDIT_BUREAU_HOUR': 'Credit_Bureau_Inquiries_Hour',
    'AMT_REQ_CREDIT_BUREAU_DAY': 'Credit_Bureau_Inquiries_Day',
    'AMT_REQ_CREDIT_BUREAU_WEEK': 'Credit_Bureau_Inquiries_Week',
    'AMT_REQ_CREDIT_BUREAU_MON': 'Credit_Bureau_Inquiries_Month',
    'AMT_REQ_CREDIT_BUREAU_QRT': 'Credit_Bureau_Inquiries_Quarter',
    'AMT_REQ_CREDIT_BUREAU_YEAR': 'Credit_Bureau_Inquiries_Year'
}

In [9]:
# Rename the columns in the train dataset
train_df.rename(columns=new_column_dict, inplace=True)

In [10]:
# Rename the columns in the test dataset
test_df.rename(columns=new_column_dict, inplace=True)

In [11]:
# View the first five rows in the test dataset
test_df.head()

Client_ID Contract_Type Gender Own_Car Own_Realty  Number_of_Children  \
0     100001    Cash loans      F       N          Y                   0   
1     100005    Cash loans      M       N          Y                   0   
2     100013    Cash loans      M       Y          Y                   0   
3     100028    Cash loans      F       N          Y                   2   
4     100038    Cash loans      M       Y          N                   1   

   Total_Income  Credit_Amount  Annuity_Amount  Goods_Price Accompanied_By  \
0      135000.0       568800.0         20560.5     450000.0  Unaccompanied   
1       99000.0       222768.0         17370.0     180000.0  Unaccompanied   
2      202500.0       663264.0         69777.0     630000.0            NaN   
3      315000.0      1575000.0         49018.5    1575000.0  Unaccompanied   
4      180000.0       625500.0         32067.0     625500.0  Unaccompanied   

  Income_Type                Education_Level Family_Status       Housing_Type  \
0     Working               Higher education       Married  House / apartment   
1     Working  Secondary / secondary special       Married  House / apartment   
2     Working               Higher education       Married  House / apartment   
3     Working  Secondary / secondary special       Married  House / apartment   
4     Working  Secondary / secondary special       Married  House / apartment   

   Region_Population_Ratio  Age_Days  Employment_Duration_Days  \
0                 0.018850    -19241                     -2329   
1                 0.035792    -18064                     -4469   
2                 0.019101    -20038                     -4458   
3                 0.026392    -13976                     -1866   
4                 0.010032    -13040                     -2191   

   Registration_Duration_Days  ID_Publish_Duration_Days  Car_Age  Has_Mobile  \
0                     -5170.0                      -812      NaN           1   
1                     -9118.0                     -1623      NaN           1   
2                     -2175.0                     -3503      5.0           1   
3                     -2000.0                     -4208      NaN           1   
4                     -4000.0                     -4262     16.0           1   

   Has_Work_Phone  Has_Work_Phone2  Has_Permanent_Mobile  Has_Phone  \
0               1                0                     1          0   
1               1                0                     1          0   
2               1                0                     1          0   
3               1                0                     1          1   
4               1                1                     1          0   

   Has_Email          Occupation  Family_Members_Count  Region_Rating  \
0          1                 NaN                   2.0              2   
1          0  Low-skill Laborers                   2.0              2   
2          0             Drivers                   2.0              2   
3          0         Sales staff                   4.0              2   
4          0                 NaN                   3.0              2   

   Region_City_Rating Application_Weekday  Application_Hour  \
0                   2             TUESDAY                18   
1                   2              FRIDAY                 9   
2                   2              MONDAY                14   
3                   2           WEDNESDAY                11   
4                   2              FRIDAY                 5   

   Different_Region_Living  Different_Region_Working  \
0                        0                         0   
1                        0                         0   
2                        0                         0   
3                        0                         0   
4                        0                         0   

   Different_Region_Living_Working  Different_City_Living  \
0                                0                      0   
1       

In [12]:
# View the first five rows in the train dataset
train_df.head()

Client_ID  Loan_Status    Contract_Type Gender Own_Car Own_Realty  \
0     100002            1       Cash loans      M       N          Y   
1     100003            0       Cash loans      F       N          N   
2     100004            0  Revolving loans      M       Y          Y   
3     100006            0       Cash loans      F       N          Y   
4     100007            0       Cash loans      M       N          Y   

   Number_of_Children  Total_Income  Credit_Amount  Annuity_Amount  \
0                   0      202500.0       406597.5         24700.5   
1                   0      270000.0      1293502.5         35698.5   
2                   0       67500.0       135000.0          6750.0   
3                   0      135000.0       312682.5         29686.5   
4                   0      121500.0       513000.0         21865.5   

   Goods_Price Accompanied_By    Income_Type                Education_Level  \
0     351000.0  Unaccompanied        Working  Secondary / secondary special   
1    1129500.0         Family  State servant               Higher education   
2     135000.0  Unaccompanied        Working  Secondary / secondary special   
3     297000.0  Unaccompanied        Working  Secondary / secondary special   
4     513000.0  Unaccompanied        Working  Secondary / secondary special   

          Family_Status       Housing_Type  Region_Population_Ratio  Age_Days  \
0  Single / not married  House / apartment                 0.018801     -9461   
1               Married  House / apartment                 0.003541    -16765   
2  Single / not married  House / apartment                 0.010032    -19046   
3        Civil marriage  House / apartment                 0.008019    -19005   
4  Single / not married  House / apartment                 0.028663    -19932   

   Employment_Duration_Days  Registration_Duration_Days  \
0                      -637                     -3648.0   
1                     -1188                     -1186.0   
2                      -225                     -4260.0   
3                     -3039                     -9833.0   
4                     -3038                     -4311.0   

   ID_Publish_Duration_Days  Car_Age  Has_Mobile  Has_Work_Phone  \
0                     -2120      NaN           1               1   
1                      -291      NaN           1               1   
2                     -2531     26.0           1               1   
3                     -2437      NaN           1               1   
4                     -3458      NaN           1               1   

   Has_Work_Phone2  Has_Permanent_Mobile  Has_Phone  Has_Email  Occupation  \
0                0                     1          1          0    Laborers   
1                0                     1          1          0  Core staff   
2                1                     1          1          0    Laborers   
3                0                     1          0          0    Laborers   
4                0                     1          0          0  Core staff   

   Family_Members_Count  Region_Rating  Region_City_Rating  \
0                   1.0              2                   2   
1                   2.0              1                   1   
2                   1.0              2                   2   
3                   2.0              2                   2   
4                   1.0              2                   2   

  Application_Weekday  Application_Hour  Different_Region_Living  \
0           WEDNESDAY                10                        0   
1              MONDAY                11                        0   
2              MONDAY                 9                        0   
3           WEDNESDAY                17                        0   
4            THURSDAY                11                        0   

   Different_Region_Working  Different_Region_Living_Working  \
0                         0                                0   
1                         0                 

I have decided to carefully study the dataset and consult recources so as to determine which features are necessary for our model, drop the ones that are not necessary and then see which missing values are left so I can proceed to handle missing values.

After careful study, the following features are needed in our data and I will be dropping the rest:

1. **Demographic Information**:
   - **Client_ID**: Unique identifier for each client, essential for tracking and analysis.
   - **Gender**: Helps understand potential demographic factors affecting loan repayment behavior.
   - **Age_Days**: Age in days provides insight into the client's age, a significant demographic factor in loan risk assessment.
   - **Family_Status**: Indicates family structure, influencing financial stability and responsibilities.

2. **Financial and Employment Details**:
   - **Total_Income**: Total income helps assess the client's financial capacity to repay loans.
   - **Employment_Duration_Days**: Duration of employment indicates stability and ability to generate income.
   - **Occupation**: Specifies the client's job type, influencing income stability and risk assessment.
   - **Organization_Type**: Identifies the industry or sector of employment, providing context on income stability.

3. **Credit and Loan Specifics**:
   - **Credit_Amount**: Amount of credit requested, indicating the financial commitment of the client.
   - **Annuity_Amount**: Monthly loan payment, critical for assessing affordability and financial strain.
   - **Goods_Price**: Price of goods purchased with the loan, providing context for credit utilization.
   - **Contract_Type**: Specifies whether the loan is a cash loan or revolving, affecting repayment structure and risk.

4. **Behavioral and Credit History**:
   - **External_Source_1, External_Source_2, External_Source_3**: Normalized scores from external sources, indicating creditworthiness and risk assessment.
   - **Credit_Bureau_Inquiries_Year**: Number of credit inquiries per year, reflecting credit-seeking behavior and potential financial stress.
   - **Obs_30_Social_Circle, Def_30_Social_Circle, Obs_60_Social_Circle, Def_60_Social_Circle**: Social circle metrics provide insights into the client's social interactions and potential financial stability.

5. **Property and Housing Details**:
   - **Own_Car, Own_Realty**: Ownership of assets indicates financial stability and commitment.
   - **Housing_Type**: Specifies the client's housing situation, influencing stability and financial priorities.

6. **Document and Communication**:
   - **Has_Document_X**: Flags indicating presence of specific documents, essential for verification and loan approval processes.
   - **Days_Since_Last_Phone_Change**: Indicates recent communication activity, reflecting client engagement and reliability.

These columns provide a comprehensive picture of the client's financial well-being.  By examining their financial health, stability, and spending habits, we gain valuable insights that help us assess the risk of loan default. This detailed analysis allows the model to predict the likelihood of repayment or default with greater accuracy, empowering lenders to make informed decisions and effectively manage risk.

In [13]:
# Select train features needed for the project
train_features_df = train_df[[
    'Client_ID',
    'Gender',
    'Age_Days',
    'Family_Status',
    'Total_Income',
    'Employment_Duration_Days',
    'Occupation',
    'Organization_Type',
    'Credit_Amount',
    'Annuity_Amount',
    'Goods_Price',
    'External_Source_1',
    'External_Source_2',
    'External_Source_3',
    'Credit_Bureau_Inquiries_Year',
    'Obs_30_Social_Circle',
    'Def_30_Social_Circle',
    'Obs_60_Social_Circle',
    'Def_60_Social_Circle',
    'Own_Car',
    'Own_Realty',
    'Housing_Type',
    'Has_Document_2',
    'Has_Document_3',
    'Has_Document_4',
    'Has_Document_5',
    'Has_Document_6',
    'Has_Document_7',
    'Has_Document_8',
    'Has_Document_9',
    'Has_Document_10',
    'Has_Document_11',
    'Has_Document_12',
    'Has_Document_13',
    'Has_Document_14',
    'Has_Document_15',
    'Has_Document_16',
    'Has_Document_17',
    'Has_Document_18',
    'Has_Document_19',
    'Has_Document_20',
    'Has_Document_21',
    'Days_Since_Last_Phone_Change',
    'Loan_Status'
]]

### **Note: The train dataset has loan status which will not be present in the test dataset, that is our target variable**

In [14]:
train_features_df.head()

Client_ID Gender  Age_Days         Family_Status  Total_Income  \
0     100002      M     -9461  Single / not married      202500.0   
1     100003      F    -16765               Married      270000.0   
2     100004      M    -19046  Single / not married       67500.0   
3     100006      F    -19005        Civil marriage      135000.0   
4     100007      M    -19932  Single / not married      121500.0   

   Employment_Duration_Days  Occupation       Organization_Type  \
0                      -637    Laborers  Business Entity Type 3   
1                     -1188  Core staff                  School   
2                      -225    Laborers              Government   
3                     -3039    Laborers  Business Entity Type 3   
4                     -3038  Core staff                Religion   

   Credit_Amount  Annuity_Amount  Goods_Price  External_Source_1  \
0       406597.5         24700.5     351000.0           0.083037   
1      1293502.5         35698.5    1129500.0           0.311267   
2       135000.0          6750.0     135000.0                NaN   
3       312682.5         29686.5     297000.0                NaN   
4       513000.0         21865.5     513000.0                NaN   

   External_Source_2  External_Source_3  Credit_Bureau_Inquiries_Year  \
0           0.262949           0.139376                           1.0   
1           0.622246                NaN                           0.0   
2           0.555912           0.729567                           0.0   
3           0.650442                NaN                           NaN   
4           0.322738                NaN                           0.0   

   Obs_30_Social_Circle  Def_30_Social_Circle  Obs_60_Social_Circle  \
0                   2.0                   2.0                   2.0   
1                   1.0                   0.0                   1.0   
2                   0.0                   0.0                   0.0   
3                   2.0                   0.0                   2.0   
4                   0.0                   0.0                   0.0   

   Def_60_Social_Circle Own_Car Own_Realty       Housing_Type  Has_Document_2  \
0                   2.0       N          Y  House / apartment               0   
1                   0.0       N          N  House / apartment               0   
2                   0.0       Y          Y  House / apartment               0   
3                   0.0       N          Y  House / apartment               0   
4                   0.0       N          Y  House / apartment               0   

   Has_Document_3  Has_Document_4  Has_Document_5  Has_Document_6  \
0               1               0               0               0   
1               1               0               0               0   
2               0               0               0               0   
3               1               0               0               0   
4               0               0               0               0   

   Has_Document_7  Has_Document_8  Has_Document_9  Has_Document_10  \
0               0               0               0                0   
1               0               0               0                0   
2               0               0               0                0   
3               0               0               0                0   
4               0               1               0                0   

   Has_Document_11  Has_Document_12  Has_Document_13  Has_Document_14  \
0                0                0                0                0   
1                0                0                0                0   
2                0                0                0                0   
3                0                0                0                0   
4                0                0                0                0   

   Has_Document_15  Has_Document_16  Has_Document_17  Has_Document_18  \
0                0                0                0                0   
1 

In [15]:
# Select test features needed for the project
test_features_df = test_df[[
    'Client_ID',
    'Gender',
    'Age_Days',
    'Family_Status',
    'Total_Income',
    'Employment_Duration_Days',
    'Occupation',
    'Organization_Type',
    'Credit_Amount',
    'Annuity_Amount',
    'Goods_Price',
    'External_Source_1',
    'External_Source_2',
    'External_Source_3',
    'Credit_Bureau_Inquiries_Year',
    'Obs_30_Social_Circle',
    'Def_30_Social_Circle',
    'Obs_60_Social_Circle',
    'Def_60_Social_Circle',
    'Own_Car',
    'Own_Realty',
    'Housing_Type',
    'Has_Document_2',
    'Has_Document_3',
    'Has_Document_4',
    'Has_Document_5',
    'Has_Document_6',
    'Has_Document_7',
    'Has_Document_8',
    'Has_Document_9',
    'Has_Document_10',
    'Has_Document_11',
    'Has_Document_12',
    'Has_Document_13',
    'Has_Document_14',
    'Has_Document_15',
    'Has_Document_16',
    'Has_Document_17',
    'Has_Document_18',
    'Has_Document_19',
    'Has_Document_20',
    'Has_Document_21',
    'Days_Since_Last_Phone_Change'
]]

In [16]:
test_features_df.head()

Client_ID Gender  Age_Days Family_Status  Total_Income  \
0     100001      F    -19241       Married      135000.0   
1     100005      M    -18064       Married       99000.0   
2     100013      M    -20038       Married      202500.0   
3     100028      F    -13976       Married      315000.0   
4     100038      M    -13040       Married      180000.0   

   Employment_Duration_Days          Occupation       Organization_Type  \
0                     -2329                 NaN            Kindergarten   
1                     -4469  Low-skill Laborers           Self-employed   
2                     -4458             Drivers       Transport: type 3   
3                     -1866         Sales staff  Business Entity Type 3   
4                     -2191                 NaN  Business Entity Type 3   

   Credit_Amount  Annuity_Amount  Goods_Price  External_Source_1  \
0       568800.0         20560.5     450000.0           0.752614   
1       222768.0         17370.0     180000.0           0.564990   
2       663264.0         69777.0     630000.0                NaN   
3      1575000.0         49018.5    1575000.0           0.525734   
4       625500.0         32067.0     625500.0           0.202145   

   External_Source_2  External_Source_3  Credit_Bureau_Inquiries_Year  \
0           0.789654           0.159520                           0.0   
1           0.291656           0.432962                           3.0   
2           0.699787           0.610991                           4.0   
3           0.509677           0.612704                           3.0   
4           0.425687                NaN                           NaN   

   Obs_30_Social_Circle  Def_30_Social_Circle  Obs_60_Social_Circle  \
0                   0.0                   0.0                   0.0   
1                   0.0                   0.0                   0.0   
2                   0.0                   0.0                   0.0   
3                   0.0                   0.0                   0.0   
4                   0.0                   0.0                   0.0   

   Def_60_Social_Circle Own_Car Own_Realty       Housing_Type  Has_Document_2  \
0                   0.0       N          Y  House / apartment               0   
1                   0.0       N          Y  House / apartment               0   
2                   0.0       Y          Y  House / apartment               0   
3                   0.0       N          Y  House / apartment               0   
4                   0.0       Y          N  House / apartment               0   

   Has_Document_3  Has_Document_4  Has_Document_5  Has_Document_6  \
0               1               0               0               0   
1               1               0               0               0   
2               0               0               0               0   
3               1               0               0               0   
4               1               0               0               0   

   Has_Document_7  Has_Document_8  Has_Document_9  Has_Document_10  \
0               0               0               0                0   
1               0               0               0                0   
2               0               1               0                0   
3               0               0               0                0   
4               0               0               0                0   

   Has_Document_11  Has_Document_12  Has_Document_13  Has_Document_14  \
0                0                0                0                0   
1                0                0                0                0   
2                0                0                0                0   
3                0                0                0                0   
4                0                0                0                0   

   Has_Document_15  Has_Document_16  Has_Document_17  Has_Document_18  \
0                0                0                0                0   
1 

    Handling Missing values

In [17]:
# Check for missing values in train set
train_missing_values = train_features_df.isna().sum()

# Print out all the results
for column, missing_count in train_missing_values.items():
    print(f"{column}: {missing_count} missing values")

Client_ID: 0 missing values
Gender: 0 missing values
Age_Days: 0 missing values
Family_Status: 0 missing values
Total_Income: 0 missing values
Employment_Duration_Days: 0 missing values
Occupation: 96391 missing values
Organization_Type: 0 missing values
Credit_Amount: 0 missing values
Annuity_Amount: 12 missing values
Goods_Price: 278 missing values
External_Source_1: 173378 missing values
External_Source_2: 660 missing values
External_Source_3: 60965 missing values
Credit_Bureau_Inquiries_Year: 41519 missing values
Obs_30_Social_Circle: 1021 missing values
Def_30_Social_Circle: 1021 missing values
Obs_60_Social_Circle: 1021 missing values
Def_60_Social_Circle: 1021 missing values
Own_Car: 0 missing values
Own_Realty: 0 missing values
Housing_Type: 0 missing values
Has_Document_2: 0 missing values
Has_Document_3: 0 missing values
Has_Document_4: 0 missing values
Has_Document_5: 0 missing values
Has_Document_6: 0 missing values
Has_Document_7: 0 missing values
Has_Document_8: 0 missing

In [18]:
# Check for missing values in test set
test_missing_values = test_features_df.isna().sum()

# Print out all the results
for column, missing_count in test_missing_values.items():
    print(f"{column}: {missing_count} missing values")

Client_ID: 0 missing values
Gender: 0 missing values
Age_Days: 0 missing values
Family_Status: 0 missing values
Total_Income: 0 missing values
Employment_Duration_Days: 0 missing values
Occupation: 15605 missing values
Organization_Type: 0 missing values
Credit_Amount: 0 missing values
Annuity_Amount: 24 missing values
Goods_Price: 0 missing values
External_Source_1: 20532 missing values
External_Source_2: 8 missing values
External_Source_3: 8668 missing values
Credit_Bureau_Inquiries_Year: 6049 missing values
Obs_30_Social_Circle: 29 missing values
Def_30_Social_Circle: 29 missing values
Obs_60_Social_Circle: 29 missing values
Def_60_Social_Circle: 29 missing values
Own_Car: 0 missing values
Own_Realty: 0 missing values
Housing_Type: 0 missing values
Has_Document_2: 0 missing values
Has_Document_3: 0 missing values
Has_Document_4: 0 missing values
Has_Document_5: 0 missing values
Has_Document_6: 0 missing values
Has_Document_7: 0 missing values
Has_Document_8: 0 missing values
Has_Doc

The two datasets have missing values across the same columns, the number of missing values might differ though. These are the columns with missing values:

Here are the column names from the provided list that have missing values:

- Occupation
- Annuity_Amount
- Goods_Price
- External_Source_1
- External_Source_2
- External_Source_3
- Credit_Bureau_Inquiries_Year
- Obs_30_Social_Circle
- Def_30_Social_Circle
- Obs_60_Social_Circle
- Def_60_Social_Circle
- Days_Since_Last_Phone_Change

Occupation: This could mean two things- either they are unemployed or its a data entry error. The observations in our train dataset is about 307,511 while the missing occupation values is 96,391. That is about 1/3rd of our dataset approximately. We can't just drop them or fill with mode so the data won't be skewed.

The goal here is to preserve the probability distribution of the occupation column and also fill missing values appropriately.

In [19]:
# Distribution of the occupation feature
occupation_probabilities_train = train_features_df['Occupation'].value_counts(normalize=True)
occupation_probabilities_test = test_features_df['Occupation'].value_counts(normalize=True)

# Combine probabilities from both training and test sets
occupation_probabilities_combined = (occupation_probabilities_train + occupation_probabilities_test) / 2

# Function to impute missing values based on combined distribution
def fill_occupation(row):
    global occupation_probabilities_combined
    if pd.isnull(row['Occupation']):
        return np.random.choice(occupation_probabilities_combined.index, p=occupation_probabilities_combined.values)
    else:
        return row['Occupation']

In [20]:
# Apply the function to impute missing values in 'Occupation'
train_features_df['Occupation'] = train_features_df.apply(fill_occupation, axis=1)
test_features_df['Occupation'] = test_features_df.apply(fill_occupation, axis=1)

Annuity_Amount: The monthly payment to be made by an individual who gets a loan, this is determined by  a number of factors such as financial capability, occupation, organization type, credit amount, housing type, family status, and maybe gender.

It would make sense to use a regression model to predict the values/entries for the 12 rows.

In [21]:
rows_with_missing_annuity = train_features_df[train_features_df['Annuity_Amount'].isna()]
rows_with_missing_annuity

Client_ID Gender  Age_Days         Family_Status  Total_Income  \
47531      155054      M    -10668  Single / not married      180000.0   
50035      157917      F     -9027        Civil marriage       94500.0   
51594      159744      F    -17376               Married      202500.0   
55025      163757      F    -11329               Married      162000.0   
59934      169487      M    -19762  Single / not married      202500.0   
75873      187985      M    -20831  Single / not married      144000.0   
89343      203726      F    -12134               Married       90000.0   
123872     243648      F    -13902        Civil marriage      202500.0   
207186     340147      M    -10151               Married      171000.0   
227939     364022      F    -16344               Married      315000.0   
239329     377174      F    -17661               Married      157500.0   
241835     379997      F    -15072               Married      315000.0   

        Employment_Duration_Days             Occupation  \
47531                      -2523  High skill tech staff   
50035                      -1270               Laborers   
51594                      -3950          Cooking staff   
55025                      -2040             Core staff   
59934                      -2498               Laborers   
75873                      -2450                Drivers   
89343                      -3721         Medicine staff   
123872                     -3540            Secretaries   
207186                      -472         Security staff   
227939                     -1478            Sales staff   
239329                     -2885               Managers   
241835                      -152            Accountants   

             Organization_Type  Credit_Amount  Annuity_Amount  Goods_Price  \
47531   Business Entity Type 3       450000.0             NaN     450000.0   
50035   Business Entity Type 1       450000.0             NaN     450000.0   
51594            Self-employed       539100.0             NaN     450000.0   
55025               Government       296280.0             NaN     225000.0   
59934                    Other       360000.0             NaN     360000.0   
75873   Business Entity Type 3       219249.0             NaN     166500.0   
89343                 Medicine       157500.0             NaN     157500.0   
123872           Self-employed       929088.0             NaN     720000.0   
207186                Security       486000.0             NaN     486000.0   
227939  Business Entity Type 3       628069.5             NaN     499500.0   
239329                   Other       792000.0             NaN     792000.0   
241835           Self-employed      1483231.5             NaN    1354500.0   

        External_Source_1  External_Source_2  External_Source_3  \
47531                 NaN           0.626896           0.372334   
50035                 NaN           0.727274           0.468660   
51594            0.885980           0.738370           0.452534   
55025            0.336803           0.566316           0.220095   
59934            0.647350           0.583947           0.177704   
75873            0.339389           0.667564           0.425893   
89343            0.682126           0.154811           0.297087   
123872           0.400723           0.510934           0.581484   
207186                NaN           0.706306           0.391055   
227939                NaN           0.618698           0.431192   
239329           0.775404           0.656305           0.651260   
241835           0.593394           0.267269                NaN   

        Credit_Bureau_Inquiries_Year  Obs_30_Social_Circle  \
47531                            1.0                   1.0   
50035                            1.0                   0.0   
51594                            1.0                   1.0   
55025                            4.0                   0.0   
59934                            6.0                   0.0   
75873             